### PR TITLE
Compute neighbor element volume on demand

### PR DESCRIPTION
**Pull request by andrsd**
MOOSE was computing the neighbor element volume even though it was not
really needed. The computation is quite expensive and shows up in
profiling.  If the volume is needed in a DG kernel, users can obtain it
by:

```
.h:
const Real & _neighbor_volume;

.C:
_neighbor_volume(getNeighborElemVolume())
```

This will tell MOOSE to compute the volume. The same holds for `InternalSideUserObject`s.

Refs #3661, #6788

### CODE REVIEW

File: framework/include/base/Assembly.h [hunk 1]
```diff
@@ -214,7 +214,7 @@ class Assembly
    * Returns the reference to the current neighbor volume
    * @return A _reference_.  Make sure to store this as a reference!
    */
-  const Real & neighborVolume() { return _current_neighbor_volume; }
+  const Real & neighborVolume() { _need_neighbor_elem_volume = true; return _current_neighbor_volume; }
```
Comment (permcody) [inline framework/include/base/Assembly.h]: I suppose this is ok. This method is probably getting big enough to move to the .C file.

**Review by permcody — APPROVED** [review]

File: framework/include/base/Assembly.h (hunk 1 above)
Comment (andrsd) [inline framework/include/base/Assembly.h]: I can move it to .C, that's not a problem...

File: framework/src/base/Assembly.C [hunk 2]
```diff
@@ -570,51 +578,53 @@ Assembly::reinitNeighbor(const Elem * neighbor, const std::vector<Point> & refer
...
-    case Moose::COORD_XYZ:
-      for (unsigned int qp = 0; qp < qrule.n_points(); qp++)
-        coord[qp] = 1.;
-      break;
-
-    case Moose::COORD_RZ:
-      for (unsigned int qp = 0; qp < qrule.n_points(); qp++)
-        coord[qp] = 2 * M_PI * q_points[qp](rz_radial_coord);
-      break;
+    FEType fe_type(neighbor->default_order(), LAGRANGE);
```
Comment (friedmud) [inline framework/src/base/Assembly.C]: This is why this is showing up in timing... there are a BUNCH of temporaries here.  These should all get moved to member variables.

File: framework/src/base/Assembly.C [hunk 3]
```diff
@@ -570,51 +578,53 @@ Assembly::reinitNeighbor(const Elem * neighbor, const std::vector<Point> & refer
...
+    std::unique_ptr<FEBase> fe(FEBase::build(neighbor->dim(), fe_type));
```
Comment (friedmud) [inline framework/src/base/Assembly.C]: This one too

File: framework/src/base/Assembly.C [hunk 4]
```diff
@@ -570,51 +578,53 @@ Assembly::reinitNeighbor(const Elem * neighbor, const std::vector<Point> & refer
...
+
+    const std::vector<Real> & JxW = fe->get_JxW();
+    const std::vector<Point> & q_points = fe->get_xyz();
+
+    // The default quadrature rule should integrate the mass matrix,
+    // thus it should be plenty to compute the area
+    QGauss qrule(neighbor->dim(), fe_type.default_quadrature_order());
```
Comment (friedmud) [inline framework/src/base/Assembly.C]: And here

File: framework/src/base/Assembly.C [hunk 5]
```diff
@@ -570,51 +578,53 @@ Assembly::reinitNeighbor(const Elem * neighbor, const std::vector<Point> & refer
...
+    fe->attach_quadrature_rule(&qrule);
+    fe->reinit(neighbor);
+
+    // set the coord transformation
+    MooseArray<Real> coord;
```
Comment (friedmud) [inline framework/src/base/Assembly.C]: And this one

**Review by friedmud — CHANGES_REQUESTED** [review]
There are a bunch of objects being dynamically allocated here...

File: framework/src/base/Assembly.C (hunk 2 above)
Comment (permcody) [inline framework/src/base/Assembly.C]: image

**Comment by friedmud** [comment]
I have no doubt it was my code!  I certainly wasn't blaming you!  I was just pointing out the real problem.  :-)

File: framework/include/base/Assembly.h [hunk 6]
```diff
@@ -657,6 +659,7 @@ class Assembly
   std::map<unsigned int, std::map<FEType, FEBase *> > _fe_neighbor;
   std::map<unsigned int, std::map<FEType, FEBase *> > _fe_face_neighbor;
   /// Each dimension's helper objects
+  std::map<unsigned int, FEBase **> _holder_fe_neighbor_helper;
```
Comment (friedmud) [inline framework/include/base/Assembly.h]: Good

File: framework/src/base/Assembly.C [hunk 7]
```diff
@@ -100,6 +100,10 @@ Assembly::Assembly(SystemBase & sys, CouplingMatrix * & cm, THREAD_ID tid) :
     (*_holder_fe_face_neighbor_helper[dim])->get_xyz();
     (*_holder_fe_face_neighbor_helper[dim])->get_JxW();
     (*_holder_fe_face_neighbor_helper[dim])->get_normals();
+
+    _holder_fe_neighbor_helper[dim] = &_fe_neighbor[dim][FEType(FIRST, LAGRANGE)];
```
Comment (friedmud) [inline framework/src/base/Assembly.C]: Looks good

**Review by friedmud — APPROVED** [review]
Looks really good.  Did you happen to do any before/after timing (not necessary - just wondering if/how much it helped)

**Comment by friedmud** [comment]
BTW: Do we have any test coverage of this?  If not - we need to add one.

**Comment by andrsd** [comment]
> BTW: Do we have any test coverage of this? If not - we need to add one.

I did not find anything that would use this, that's why I made it into a subscription based API. If nobody needs it, let's not spend time doing it at all ;-)

Not sure how we could test it, nothing obvious comes to mind. A unit test would be good, but probably quite a bit of code to actually get down to that call...

**Comment by andrsd** [comment]
> Did you happen to do any before/after timing

Some timing is already [here](https://github.com/idaholab/moose/issues/6788#issuecomment-256176659), I will do one more with the neighbor element volume computation enabled with this latest optimization...

**Comment by friedmud** [comment]
What kind of object is triggering that whole function right now?  Is it
Mortar stuff?

**Comment by andrsd** [comment]
> What kind of object is triggering that whole function right now?

Nothing in the MOOSE universe needs the computation of neighbor volume. That's why the app test passed without patching anything. The computation is off by default and when something needs it, they can enable it via an API call. That was the primary purpose of this PR.

**Comment by permcody** [comment]
@friedmud - I'm satisfied with this PR. Feel free to merge it, otherwise I'll get to it later today.

**Pull request merged**